packages/flutter/lib/src/widgets/sensitive_content.dart

```diff
@@ -0,0 +1,225 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart' show ContentSensitivity, SensitiveContentService;
+
+import 'framework.dart';
+
+/// Data structure used to track the number of [SensitiveContent] widgets with
+/// each of the different [ContentSensitivity] levels set in a particular
+/// Flutter view.
+class ViewContentSensitivityState {
+  /// Creates a [ViewContentSensitivityState].
+  ViewContentSensitivityState();
+
+  /// The current [ContentSensitivity] level set for the Flutter view that this
+  /// state represents.
+  /// 
+  /// By default, this level is [ContentSensitivity.autoSensitive] because this
+  /// is the default level on Android and this feature is currently only
+  /// supported for Android.
+  ContentSensitivity currentContentSensitivitySetting = ContentSensitivity.autoSensitive;
+
+  /// A map that contains the number of [SensitiveContent] widgets that have
+  /// each of the different [ContentSensitivity] levels.
+  final Map<ContentSensitivity, int> contentSensitivityCounts = <ContentSensitivity, int> {
+    ContentSensitivity.sensitive: 0,
+    ContentSensitivity.autoSensitive: 0,
+    ContentSensitivity.notSensitive: 0,
+  };
+
+  /// Increases the count of [SensitiveContent] widgets with [sensitivityLevel] set.
+  void addWidgetWithContentSensitivity(ContentSensitivity sensitivityLevel) {
+    contentSensitivityCounts[sensitivityLevel] = contentSensitivityCounts[sensitivityLevel]! + 1;
+  }
+
+  /// Decreases the count of [SensitiveContent] widgets with [sensitivityLevel] set.
+  void removeWidgetWithContentSensitivity(ContentSensitivity sensitivityLevel) { 
+    contentSensitivityCounts[sensitivityLevel] = contentSensitivityCounts[sensitivityLevel]! - 1;
+  }
+
+  /// Returns the number of [SenstiveContent] widgets represented by this state.
+  int getTotalNumberOfWidgets() {
+    return contentSensitivityCounts.values.reduce((int sum, int value) => sum + value);
+  }
+}
+
+/// Host of the current content sensitivity level for each Flutter view that
+/// contains any [SensitiveContent] widgets.
+class SensitiveContentSetting {
+  SensitiveContentSetting._();
+
+  final Map<int, ViewContentSensitivityState> _contentSensitivityStates = <int, ViewContentSensitivityState> {};
+  final SensitiveContentService _sensitiveContentService = SensitiveContentService();
+
+  static final SensitiveContentSetting _instance = SensitiveContentSetting._();
```

Comment:
(I thought I left a comment about this elsewhere but can't find it, maybe it got deleted.)

I know this is how we already do this all over the framework, but I wish we could avoid this singleton pattern. The problem is that it won't reset itself between test runs or hot restarts. I wonder if we could instantiate this in WidgetsApp or somewhere like that? I probably should poke around and see if I can find an example of something like that.

Reply:
This is the comment I was thinking of, we can focus the conversation there: https://github.com/flutter/flutter/pull/158473/files#r1909479844